### PR TITLE
Add support for v1 of API

### PR DESCRIPTION
**Pull request by tmiddlet2666**
Implement new gRPC v1 client in coherence server.

### CODE REVIEW

File: coherence/session.go [hunk 1]
```diff
@@ -349,8 +431,8 @@ func (s *Session) Close() {
 }
 
 func (s *Session) String() string {
-	return fmt.Sprintf("Session{id=%s, closed=%v, caches=%d, maps=%d, options=%v}", s.sessionID.String(), s.closed,
-		len(s.caches), len(s.maps), s.sessOpts)
+	return fmt.Sprintf("Session{id=%s, closed=%v, caches=%d, maps=%d, gRPCv1=%v, options=%v}", s.sessionID.String(), s.closed,
```
Comment (rlubke) [inline coherence/session.go:434]: Maybe gRPCv1 should be protocol-version.  Looking forward to the next rev of the protocol, it doesn't seem to make sense to explicitly list the version in the attribute name.

File: coherence/session.go [hunk 2]
```diff
@@ -380,7 +462,7 @@ func (s *Session) ensureConnection() error {
 			if s.GetReadyTimeout() != 0 {
 				return waitForReady(s)
 			}
-			s.debug(fmt.Sprintf("session: %s attempting connection to address %s", s.sessionID, s.sessOpts.Address))
+			log.Printf("INFO: Session: [%s] attempting connection to address %s", s.sessionID, s.sessOpts.Address)
```
Comment (rlubke) [inline coherence/session.go:465]: Should this be using logMessage(INFO, ...) ?

File: coherence/session.go [hunk 3]
```diff
@@ -415,14 +497,28 @@ func (s *Session) ensureConnection() error {
 	conn, err := grpc.DialContext(newCtx, s.sessOpts.Address, s.dialOptions...)
 
 	if err != nil {
-		log.Printf("could not connect. Reason: %v", err)
+		log.Printf("WARN: Session: [%s] could not connect. Reason: %v", s.sessionID, err)
```
Comment (rlubke) [inline coherence/session.go:500]: Should this be using logMessage(WARN, ...) ?

File: coherence/session.go [hunk 4]
```diff
@@ -415,14 +497,28 @@ func (s *Session) ensureConnection() error {
...
+	manager, err1 := newStreamManagerV1(s, cacheServiceProtocol)
+	if err1 == nil {
+		// save the stream manager for a successful V1 client connection
+		s.v1StreamManagerCache = manager
+		apiMessage = fmt.Sprintf(", %v", manager)
+	} else {
+		s.debug("error connecting to session via v1, falling back to v0: %v", err1)
+	}
+
+	log.Printf("INFO: Session [%s] connected to [%s]%s", s.sessionID, s.sessOpts.Address, apiMessage)
```
Comment (rlubke) [inline coherence/session.go:519]: Should this be using logMessage(INFO, ...) ?

File: coherence/session.go [hunk 5]
```diff
@@ -530,7 +638,8 @@ func waitForReady(s *Session) error {
 			return nil
 		}
 		if !messageLogged {
-			log.Printf("State is %v, waiting until ready timeout of %v for valid connection", state, readyTimeout)
+			log.Printf("INFO: Session [%s] State is %v, waiting until ready timeout of %v for valid connection",
```
Comment (rlubke) [inline coherence/session.go:641]: Should this be using logMessage(INFO, ...) ?

File: coherence/v1client.go [hunk 6]
```diff
@@ -0,0 +1,1371 @@
...
+		return false, err1
+	}
+
+	return unwrapBool(result)
+}
+
+// remove issues a get request for a given key and returns the bytes value which could be nil.
+func (m *streamManagerV1) remove(ctx context.Context, cache string, key []byte) (*[]byte, error) {
+	req, err := m.newRemoveRequest(cache, key)
+	if err != nil {
```
Comment (rlubke) [inline coherence/v1client.go:601]: The code from 601 to 621 seems to be repeated quite a bit.  Could this be abstracted?

File: coherence/v1client.go [hunk 7]
```diff
@@ -0,0 +1,1371 @@
...
+
+	// create a channel that will be populated by the streaming request
+	ch := make(chan BinaryKeyAndValue)
+
+	newCtx, cancel := m.session.ensureContext(ctx)
+
+	// this channel will receive entries back from the stream
+	respChannel := waitForStreamingResponse(newCtx, requestType.ch)
+
+	go func() {
```
Comment (rlubke) [inline coherence/v1client.go:975]: This loop is duplicated several times.  Could it be generalized such that you can pass function taking the in coming message and the returning the unpacked response?

File: coherence/v1requests_test.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ */
+
+package coherence
+
+import (
+	"github.com/onsi/gomega"
+	"testing"
+)
+
+func TestNewNamedCacheRequest(t *testing.T) {
```
Comment (rlubke) [inline coherence/v1requests_test.go:14]: What is this test actually doing?

File: coherence/session.go [hunk 9]
```diff
@@ -504,6 +607,11 @@ func (s *Session) ensureConnection() error {
 	return nil
 }
 
+// IsGrpcV1 indicates if we are connected to a v1 gRPC client.
+func (s *Session) IsGrpcV1() bool {
```
Comment (rlubke) [inline coherence/session.go:611]: I think a more general call would be better here.  The protocol version could rev in the future.  Would it be better to have an accessor for the protocol version vs a one with the protocol version hard-coded?

**Review by rlubke — DISMISSED** [review]

File: coherence/session.go (hunk 1 above)
Comment (tmiddlet2666) [reply coherence/session.go:434]: sounds good i will add serverProtocolVersion and rename 
`func (s *Session) IsGrpcV1` to `func (s *Session) IsGrpcV1OrAbove`

File: coherence/v1client.go (hunk 6 above)
Comment (tmiddlet2666) [reply coherence/v1client.go:601]: good catch have made generic function to replace this

File: coherence/v1client.go (hunk 7 above)
Comment (tmiddlet2666) [reply coherence/v1client.go:975]: As discussed, the de-duplication of this is quite complex as they return different types and the initial stab at the code makes it much more complex to maintain.

**Review by rlubke — APPROVED** [review]

File: coherence/v1requests_test.go (hunk 8 above)
Comment (tmiddlet2666) [reply coherence/v1requests_test.go]: This was a mistake and should have been deleted. Done

**Pull request merged**